// angular-oauth2-oidc/esm2015/token-validation/hash-handler.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { sha256 } from './js-sha256';
/**
 * Abstraction for crypto algorithms
 */
import * as ɵngcc0 from '@angular/core';
export class HashHandler {
}
export class DefaultHashHandler {
    calcHash(valueToHash, algorithm) {
        return __awaiter(this, void 0, void 0, function* () {
            // const encoder = new TextEncoder();
            // const hashArray = await window.crypto.subtle.digest(algorithm, data);
            // const data = encoder.encode(valueToHash);
            const hashArray = sha256.array(valueToHash);
            // const hashString = this.toHashString(hashArray);
            const hashString = this.toHashString2(hashArray);
            return hashString;
        });
    }
    toHashString2(byteArray) {
        let result = '';
        for (let e of byteArray) {
            result += String.fromCharCode(e);
        }
        return result;
    }
    toHashString(buffer) {
        const byteArray = new Uint8Array(buffer);
        let result = '';
        for (let e of byteArray) {
            result += String.fromCharCode(e);
        }
        return result;
    }
}
DefaultHashHandler.ɵfac = function DefaultHashHandler_Factory(t) { return new (t || DefaultHashHandler)(); };
DefaultHashHandler.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: DefaultHashHandler, factory: DefaultHashHandler.ɵfac });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DefaultHashHandler, [{
        type: Injectable
    }], null, null); })();
